### PR TITLE
Support ScatterOp Codegen

### PR DESCRIPTION
**Pull request by eedalong**

### CODE REVIEW

File: tao_compiler/mlir/disc/IR/lhlo_disc_ops.td [hunk 1]
```diff
@@ -295,5 +295,26 @@ def LHLO_ArgsMutationOp : LHLODISC_Op<"args_mutation", []> {
   );
 }
 
+def LHLODISC_InplaceScatterOp: LHLODISC_Op<"inplace_scatter", []> {
```
Comment (Yancey1989) [inline tao_compiler/mlir/disc/IR/lhlo_disc_ops.td:298]: do we need a new operator to define a  **INPLACE** op? Maybe we simple way is `if (op.operand[0] == op.operand[-1])` ?

File: tao_compiler/mlir/disc/transforms/tests/scatter-legalize.mlir [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+// CHECK-SAME: %[[ARG0:.*]]: memref<32000x4096xf32>, %[[ARG1:.*]]: memref<8193x1xi64>, %[[ARG2:.*]]: memref<8192x4096xf32>
+func.func @test_scatterop_lowering(%arg0: tensor<32000x4096xf32>, %arg1: tensor<8192x1xi64>, %arg2: tensor<8192x4096xf32>) -> tensor<32000x4096xf32>
+  attributes {
+    tf.entry_function = {
+      input_placements = "cpu, cpu, cpu",
+      inputs = "input0, input1, input2",
+      output_placements = "cpu",
+      outputs = "output0"
+    }} {
+    %2 = "mhlo.scatter"(%arg0, %arg1, %arg2) ({
```
Comment (Yancey1989) [inline tao_compiler/mlir/disc/transforms/tests/scatter-legalize.mlir:17]: this ut is too simple, please check the IR of scatter op lowing.

File: pytorch_blade/pytorch_blade/torch-mlir/lib/Conversion/TorchToMhlo/DiscTorchToMhlo.cpp [hunk 3]
```diff
@@ -1590,6 +1591,219 @@ LogicalResult ConvertAtenOp<AtenSliceScatterOp>::matchAndRewrite(
...
+  }
+
+  Value baseIndicesValue = rewriter.create<tensor::FromElementsOp>(loc, baseIndices);
+  Value stridesValue = rewriter.create<tensor::FromElementsOp>(loc, strides);
+  Value limitIndicesValue = rewriter.create<tensor::FromElementsOp>(loc, limitIndices);
+
+  auto sliceOpResultType = RankedTensorType::get(indexType.getShape(), srcType.getElementType());
+  src = rewriter.create<mhlo::RealDynamicSliceOp>(loc, getTypeConverter()->convertType(sliceOpResultType), src, baseIndicesValue, limitIndicesValue, stridesValue);
+
+  // Construct ScatterDimensionNumbersAttr
```
[the remark this answers is not in the document]
Comment (eedalong) [reply pytorch_blade/pytorch_blade/torch-mlir/lib/Conversion/TorchToMhlo/DiscTorchToMhlo.cpp:1644]: Fixed

File: tao_compiler/mlir/disc/IR/lhlo_disc_ops.td (hunk 1 above)
Comment (eedalong) [reply tao_compiler/mlir/disc/IR/lhlo_disc_ops.td]: Yeah you are right. Fixed.

File: tao_compiler/mlir/disc/transforms/tests/scatter-legalize.mlir (hunk 2 above)
Comment (eedalong) [reply tao_compiler/mlir/disc/transforms/tests/scatter-legalize.mlir:10]: Fixed.

**Pull request closed without merging**